node/bft/src/helpers/storage.rs

```diff
@@ -1131,6 +1131,213 @@ pub(crate) mod tests {
+        }
+    }
+
+    /// Make sure that we reject all certificates without sufficient signatures early.
+    #[test]
+    fn test_invalid_incoming_certificate_missing_signature() {
+        let rng = &mut TestRng::default();
+
+        // Sample a committee.
+        let (committee, private_keys) =
+            snarkvm::ledger::committee::test_helpers::sample_committee_and_keys_for_round(1, 10, rng);
+        // Initialize the ledger.
+        let ledger = Arc::new(MockLedgerService::new(committee));
+        // Initialize the storage.
+        let storage = Storage::<CurrentNetwork>::new(ledger, Arc::new(BFTMemoryService::new()), 1);
+
+        // Go through many rounds of valid certificates and ensure they're accepted.
+        let mut previous_certs = IndexSet::default();
+
+        for round in 1..=5 {
+            let mut new_certs = IndexSet::default();
+
+            // Generate one cert per validator
+            for private_key in private_keys.iter() {
+                if round < 5 {
+                    let other_keys: Vec<_> = private_keys.iter().cloned().filter(|k| k != private_key).collect();
+
+                    let certificate = sample_batch_certificate_for_round_with_committee(
+                        round,
+                        previous_certs.clone(),
+                        private_key,
+                        &other_keys,
+                        rng,
+                    );
+                    storage.check_incoming_certificate(&certificate).expect("Valid certificate rejected");
+                    new_certs.insert(certificate.id());
+
+                    // Construct the sample 'transmissions'.
+                    let (missing_transmissions, _transmissions) = sample_transmissions(&certificate, rng);
+                    storage.insert_certificate_atomic(certificate, Default::default(), missing_transmissions);
+                } else {
+                    // Pick a few signers, but not enough to form a quorum.
+                    let other_keys: Vec<_> = private_keys[0..=3].iter().cloned().filter(|k| k != private_key).collect();
+
+                    let certificate = sample_batch_certificate_for_round_with_committee(
+                        round,
+                        previous_certs.clone(),
+                        private_key,
+                        &other_keys,
+                        rng,
+                    );
+                    assert!(storage.check_incoming_certificate(&certificate).is_err());
+                }
+            }
+
+            previous_certs = new_certs;
+        }
+    }
+
+    /// Verify that `insert_certificate` rejects certs with less edges than required.
+    #[test]
+    fn test_invalid_certificate_insufficient_previous_certs() {
+        let rng = &mut TestRng::default();
+
+        // Sample a committee.
+        let (committee, private_keys) =
+            snarkvm::ledger::committee::test_helpers::sample_committee_and_keys_for_round(1, 10, rng);
```

Comment:
Same comment as above here, about 0 instead of 1 as the committee starting round.